numba/runtime/_nrt_python.c

```diff
@@ -0,0 +1,539 @@
+
+    if (!PyArray_Check(obj)) {
+        return -1;
+    }
+
+    ndary = (PyArrayObject*)obj;
+    ndim = PyArray_NDIM(ndary);
+    data = PyArray_DATA(ndary);
+
+    arystruct->meminfo = meminfo_new_from_pyobject((void*)data, obj);
+    arystruct->data = data;
+    arystruct->nitems = PyArray_SIZE(ndary);
+    arystruct->itemsize = PyArray_ITEMSIZE(ndary);
+    arystruct->parent = obj;
+    p = arystruct->shape_and_strides;
+    for (i = 0; i < ndim; i++, p++) {
+        *p = PyArray_DIM(ndary, i);
+    }
+    for (i = 0; i < ndim; i++, p++) {
+        *p = PyArray_STRIDE(ndary, i);
+    }
+
+    NRT_Debug(nrt_debug_print("NRT_adapt_ndarray %p\n", arystruct->meminfo));
+
+    NRT_MemInfo_acquire(arystruct->meminfo);
+    return 0;
+}
+
+static
+PyObject* try_to_return_parent(arystruct_t *arystruct, int ndim, int typenum)
+{
+    int i;
+    PyArrayObject *array = (PyArrayObject *)arystruct->parent;
+
+    if (PyArray_DATA(array) != arystruct->data)
+        goto RETURN_ARRAY_COPY;
+
+    if (PyArray_TYPE(array) != typenum)
+        goto RETURN_ARRAY_COPY;
+
+    if (PyArray_NDIM(array) != ndim)
+        goto RETURN_ARRAY_COPY;
+
+    if (PyArray_ITEMSIZE(array) != arystruct->itemsize)
+        goto RETURN_ARRAY_COPY;
+
+    for(i = 0; i < ndim; ++i) {
+        if (PyArray_DIMS(array)[i] != arystruct->shape_and_strides[i])
+            goto RETURN_ARRAY_COPY;
+        if (PyArray_STRIDES(array)[i] != arystruct->shape_and_strides[ndim + i])
+            goto RETURN_ARRAY_COPY;
+    }
+
+    /* Yes, it is the same array
+       Return new reference */
+    Py_INCREF((PyObject *)array);
+    return (PyObject *)array;
+
+RETURN_ARRAY_COPY:
+    return NULL;
+
+}
+
+static
+PyObject* NRT_adapt_native_array(arystruct_t* arystruct, int ndim,
+                                 int type_num) {
+    PyObject *array;
+    MemInfoObject *miobj = NULL;
+    PyObject *args;
+    npy_intp *shape, *strides;
+    int flags=0;
+
+    if (arystruct->parent) {
+        array = try_to_return_parent(arystruct, ndim, type_num);
+        if (array) return array;
+    }
+
+    if (arystruct->meminfo) {
+        /* wrap into MemInfoObject */
+        miobj = PyObject_New(MemInfoObject, &MemInfoType);
+        args = Py_BuildValue("(K)", (unsigned PY_LONG_LONG)arystruct->meminfo);
+        if(MemInfo_init(miobj, args, NULL)) {
+            return NULL;
+        }
+        Py_DECREF(args);
+        /* Set writable */
+        flags |= NPY_ARRAY_WRITEABLE;
+    }
+
+    shape = arystruct->shape_and_strides;
+    strides = shape + ndim;
+    array = PyArray_New(&PyArray_Type, ndim, shape, type_num,
+                        strides, arystruct->data,
+                        arystruct->itemsize, flags, (PyObject*)miobj);
+
+    if (miobj) {
+        /* Set the MemInfoObject as the base object */
+        PyArray_SetBaseObject((PyArrayObject*)array, (PyObject *)miobj);
+    }
+    return array;
+}
+
+static void
+NRT_adapt_buffer(Py_buffer *buf, arystruct_t *arystruct)
```

Comment:
pycc is not currently useful (it's functionally was fundamentally broken in the Great Numba Refactor), but we do get requests from people for the ability to compile code such that the Numba module is no longer needed.

We have several things to resolve before that is possible, but I think we do want to plan for how nopython mode code can be made independent of Numba.